File: 09-CustomDialog-VB/Dialogs/TablePropertiesDialog.vb
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Globalization
Imports System.Windows.Forms
Imports SpiceLogic.HtmlEditor.Abstractions.Dialogs
Imports SpiceLogic.HtmlEditor.Abstractions.Entities
Imports SpiceLogic.HtmlEditor.WinForms.Helpers
Imports SpiceLogic.HtmlEditor.WinForms.Models.Dialogs

Namespace Global.CustomDialog.Dialogs

    ''' <summary>
    ''' Class TablePropertiesDialog
    ''' </summary>
    Partial Public Class TablePropertiesDialog
        Inherits DialogBase
        Implements ITableDialog

        ''' <summary>
        ''' The _element
        ''' </summary>
        Private _element As TableElement
        ''' <summary>
        ''' The _table cell dialog
        ''' </summary>
        Private ReadOnly _tableCellDialog As ITableCellDialog
        ''' <summary>
        ''' The _background picture URL
        ''' </summary>
        Private _backgroundPictureUrl As String = String.Empty

        Private ReadOnly _propertiesAffected As List(Of String) =
            New List(Of String) From {
            "Rows",
            "Columns",
            "Width",
            "Height",
            "Caption",
            "BorderWidth",
            "CellPadding",
            "CellSpacing",
            "BorderColor",
            "BorderAttr",
            "BorderStyle",
            "BgColor",
            "BorderCollapse",
            "SummaryDescription",
            "ID",
            "Name",
            "CSSClassName",
            "BackGround"
        }

        ''' <summary>
        ''' Initializes a new instance of the <see cref="TablePropertiesDialog" /> class.
        ''' </summary>
        ''' <param name="cellDialog">The cell dialog.</param>
        Public Sub New(cellDialog As ITableCellDialog)
            InitializeComponent()
            ' Spec 019 FR-009: cmbBorderStyle.SelectedIndex = 0 moved to the
            ' .Designer.vb file -- design-time serialization, no runtime override.
            Me._tableCellDialog = cellDialog
            Me.CancelButton = btnCancel

            Icon = Constants.FormIcon
        End Sub

        ''' <summary>
        ''' Handles the Load event of the TablePropertiesDialog control.
        ''' </summary>
        ''' <param name="sender">The source of the event.</param>
        ''' <param name="e">The <see cref="EventArgs"/> instance containing the event data.</param>
        Private Sub TablePropertiesDialog_Load(sender As Object, e As EventArgs) Handles MyBase.Load
            btnCellProperties.Visible = Me.getOrInitCellElement() IsNot Nothing
        End Sub

        ''' <summary>
        ''' Gets or sets the element.
        ''' </summary>
        ''' <value>The element.</value>
        Public Property Element As TableElement Implements ITableDialog.Element
            Get
                If _element Is Nothing Then
                    _element = New TableElement(Nothing)
                End If

                _element.ResetValues(Me._propertiesAffected)

                _element.Rows = CInt(numRows.Value)
                _element.Columns = CInt(numCols.Value)

                If chkCellPadding.Checked Then
                    _element.CellPadding = CInt(numCellPadding.Value)
                End If
                If chkCellSpacing.Checked Then
                    _element.CellSpacing = CInt(numCellSpacing.Value)
                End If

                If chkWidth.Checked AndAlso cmbWidthUnit.SelectedItem IsNot Nothing Then
                    _element.Width = Convert.ToInt32(txtWidth.Text.Trim())
                    _element.WidthUnit = cmbWidthUnit.SelectedItem.ToString()
                End If

                If chkHeight.Checked AndAlso cmbHeightUnit.SelectedItem IsNot Nothing Then
                    _element.Height = Convert.ToInt32(txtHeight.Text.Trim())
                    _element.HeightUnit = cmbHeightUnit.SelectedItem.ToString()
                End If

                If chkCaption.Checked Then
                    _element.Caption = txtCaption.Text.Trim()
                End If

                If chkBorderWidth.Checked Then
                    If chkBorderStyle.Checked AndAlso cmbBorderStyle.SelectedItem IsNot Nothing Then
                        _element.BorderStyle = cmbBorderStyle.SelectedItem.ToString()
                        _element.BorderWidth = CInt(numBorderWidth.Value)
                    Else
                        _element.BorderAttr = CInt(numBorderWidth.Value)
                    End If
                End If

                If chklnkBorderColor.Enabled AndAlso chklnkBorderColor.Checked Then
                    _element.BorderColor = ColorTranslator.ToHtml(txtBorderColor.BackColor)
                End If

                If chklnkBgColor.Checked Then
                    _element.BgColor = ColorTranslator.ToHtml(txtBgColor.BackColor)
                End If

                _element.SummaryDescription = txtSummaryDescription.Text
                _element.BorderCollapse = chkBorderCollapse.Enabled AndAlso chkBorderCollapse.Checked
                _element.Id = txtId.Text.Trim()
                _element.Name = txtName.Text.Trim()
                _element.CssClassName = txtClassName.Text.Trim()
                _element.CssStyle = txtCss.Text.Trim()

                If chklnkBackgroundPicture.Checked Then
                    _element.BackGround = pbBackgroundPicture.ImageLocation
                End If

                _element.BorderToAll = chkBorderToAll.Checked

                Return _element
            End Get
            Set(value As TableElement)
                _element = value
                If _element Is Nothing Then
                    Return
                End If

                ' Spec 019 FR-009: every element→UI assignment is gated on the
                ' element carrying data, so design-time defaults survive otherwise.
                If _element.Rows > 0 Then
                    numRows.Value = _element.Rows
                End If
                If _element.Columns > 0 Then
                    numCols.Value = _element.Columns
                End If

                If _element.Width.HasValue Then
                    chkWidth.Checked = True
                    txtWidth.Text = _element.Width.Value.ToString(CultureInfo.InvariantCulture)
                    If _element.WidthUnit IsNot Nothing Then
                        cmbWidthUnit.SelectedItem = _element.WidthUnit
                    End If
                End If

                If _element.Height.HasValue Then
                    chkHeight.Checked = True
                    txtHeight.Text = _element.Height.Value.ToString(CultureInfo.InvariantCulture)
                    If _element.HeightUnit IsNot Nothing Then
                        cmbHeightUnit.SelectedItem = _element.HeightUnit
                    End If
                End If

                If Not String.IsNullOrEmpty(_element.Caption) Then
                    chkCaption.Checked = True
                    txtCaption.Text = _element.Caption
                End If

                If _element.BorderWidth.HasValue OrElse _element.BorderAttr.HasValue Then
                    chkBorderWidth.Checked = True
                End If

                If Not String.IsNullOrEmpty(_element.BorderStyle) Then
                    chkBorderStyle.Checked = True
                    cmbBorderStyle.Enabled = True
                    chkBorderCollapse.Enabled = True
                    chklnkBorderColor.Enabled = True
                    Me.txtBorderColor.Enabled = Me.chklnkBorderColor.Checked
                    cmbBorderStyle.Text = _element.BorderStyle
                    If chkBorderWidth.Checked Then
                        numBorderWidth.Value = If(_element.BorderWidth, 0)
                    End If
                ElseIf chkBorderWidth.Checked Then
                    numBorderWidth.Value = If(_element.BorderAttr, If(_element.BorderWidth, 0))
                End If

                If _element.CellPadding.HasValue Then
                    chkCellPadding.Checked = True
                    numCellPadding.Value = _element.CellPadding.Value
                End If

                If _element.CellSpacing.HasValue Then
                    chkCellSpacing.Checked = True
                    numCellSpacing.Value = _element.CellSpacing.Value
                End If

                If Not String.IsNullOrEmpty(_element.BorderColor) Then
                    chklnkBorderColor.Checked = True
                    Dim converter As ColorConverter = New ColorConverter()
                    txtBorderColor.BackColor = CType(converter.ConvertFromString(_element.BorderColor), Color)
                End If

                If Not String.IsNullOrEmpty(_element.BgColor) Then
                    chklnkBgColor.Checked = True
                    Dim converter As ColorConverter = New ColorConverter()
                    txtBgColor.BackColor = CType(converter.ConvertFromString(_element.BgColor), Color)
                End If

                If _element.BorderCollapse IsNot Nothing AndAlso _element.BorderCollapse.Value Then
                    chkBorderCollapse.Checked = True
                End If
                ' Spec 019 FR-009: preserve design-time defaults on empty fields.
                If Not String.IsNullOrEmpty(_element.SummaryDescription) Then
                    txtSummaryDescription.Text = _element.SummaryDescription
                End If
                If Not String.IsNullOrEmpty(_element.Id) Then
                    txtId.Text = _element.Id
                End If
                If Not String.IsNullOrEmpty(_element.Name) Then
                    txtName.Text = _element.Name
                End If
                If Not String.IsNullOrEmpty(_element.CssClassName) Then
                    txtClassName.Text = _element.CssClassName
                End If
                '///////////////
                If Not String.IsNullOrEmpty(_element.BackGround) Then
                    chklnkBackgroundPicture.Checked = True
                    pbBackgroundPicture.ImageLocation = _element.BackGround
                    pbBackgroundPicture.Load()
                End If

                Dim cssText As String = _element.GetCssStyleWithoutProperties(Me._propertiesAffected)
                If Not String.IsNullOrEmpty(cssText) Then
                    txtCss.Text = cssText
                End If
                chkBorderToAll.Checked = True
            End Set
        End Property

#Region "------------ UI Event Handlers --------------"

        ''' <summary>
        ''' Handles the CheckedChanged event of the chkWidth control.
        ''' </summary>
        ''' <param name="sender">The source of the event.</param>
        ''' <param name="e">The <see cref="System.EventArgs" /> instance containing the event data.</param>
        Private Sub chkWidth_CheckedChanged(sender As Object, e As EventArgs) Handles chkWidth.CheckedChanged
            txtWidth.Enabled = chkWidth.Checked
            cmbWidthUnit.Enabled = chkWidth.Checked
        End Sub

        ''' <summary>
        ''' Handles the CheckedChanged event of the chkHeight control.
        ''' </summary>
        ''' <param name="sender">The source of the event.</param>
        ''' <param name="e">The <see cref="System.EventArgs" /> instance containing the event data.</param>
        Private Sub chkHeight_CheckedChanged(sender As Object, e As EventArgs) Handles chkHeight.CheckedChanged
            txtHeight.Enabled = chkHeight.Checked
            cmbHeightUnit.Enabled = chkHeight.Checked
        End Sub

        ''' <summary>
        ''' Handles the CheckedChanged event of the chkBorderWidth control.
        ''' </summary>
        ''' <param name="sender">The source of the event.</param>
        ''' <param name="e">The <see cref="System.EventArgs" /> instance containing the event data.</param>
        Private Sub chkBorderWidth_CheckedChanged(sender As Object, e As EventArgs) Handles chkBorderWidth.CheckedChanged
            numBorderWidth.Enabled = chkBorderWidth.Checked

            Me.chkBorderStyle.Enabled = Me.chkBorderWidth.Checked

            '            chkBorderStyle.Checked = chkBorderWidth.Checked;
            cmbBorderStyle.Enabled = chkBorderWidth.Checked AndAlso chkBorderStyle.Checked
            If Not chkBorderWidth.Checked Then chkBorderToAll.Checked = True
        End Sub

        ''' <summary>
        ''' Handles the CheckedChanged event of the chkCellPadding control.
        ''' </summary>
        ''' <param name="sender">The source of the event.</param>
        ''' <param name="e">The <see cref="System.EventArgs" /> instance containing the event data.</param>
        Private Sub chkCellPadding_CheckedChanged(sender As Object, e As EventArgs) Handles chkCellPadding.CheckedChanged
            numCellPadding.Enabled = chkCellPadding.Checked
        End Sub

        ''' <summary>
        ''' Handles the CheckedChanged event of the chkCellSpacing control.
        ''' </summary>
        ''' <param name="sender">The source of the event.</param>
        ''' <param name="e">The <see cref="System.EventArgs" /> instance containing the event data.</param>
        Private Sub chkCellSpacing_CheckedChanged(sender As Object, e As EventArgs) Handles chkCellSpacing.CheckedChanged
            numCellSpacing.Enabled = chkCellSpacing.Checked
        End Sub

        ''' <summary>
        ''' Handles the LinkClicked event of the chklnkBorderColor control.
        ''' </summary>
        ''' <param name="sender">The source of the event.</param>
        ''' <param name="e">The <see cref="System.Windows.Forms.LinkLabelLinkClickedEventArgs" /> instance containing the event data.</param>
        Private Sub chklnkBorderColor_LinkClicked(sender As Object, e As LinkLabelLinkClickedEventArgs) Handles chklnkBorderColor.LinkClicked
            Using myDialog As ColorDialog = New ColorDialog() With {
                .AllowFullOpen = True,
                .AnyColor = True,
                .Color = txtBorderColor.BackColor,
                .CustomColors = New Integer() {ColorTranslator.ToOle(txtBorderColor.BackColor)}
            }
                If myDialog.ShowDialog() = DialogResult.OK Then
                    txtBorderColor.BackColor = myDialog.Color
                End If
            End Using
        End Sub

        ''' <summary>
        ''' Handles the LinkClicked event of the chklnkBgColor control.
        ''' </summary>
        ''' <param name="sender">The source of the event.</param>
        ''' <param name="e">The <see cref="System.Windows.Forms.LinkLabelLinkClickedEventArgs" /> instance containing the event data.</param>
        Private Sub chklnkBgColor_LinkClicked(sender As Object, e As LinkLabelLinkClickedEventArgs) Handles chklnkBgColor.LinkClicked
            Using myDialog As ColorDialog = New ColorDialog() With {
                .AllowFullOpen = True,
                .AnyColor = True,
                .Color = txtBgColor.BackColor,
                .CustomColors = New Integer() {ColorTranslator.ToOle(txtBgColor.BackColor)}
            }
                If myDialog.ShowDialog() = DialogResult.OK Then
                    txtBgColor.BackColor = myDialog.Color
                End If
            End Using
        End Sub

        ''' <summary>
        ''' Handles the LinkClicked event of the chklnkBackgroundPicture control.
        ''' </summary>
        ''' <param name="sender">The source of the event.</param>
        ''' <param name="e">The <see cref="System.Windows.Forms.LinkLabelLinkClickedEventArgs" /> instance containing the event data.</param>
        Private Sub chklnkBackgroundPicture_LinkClicked(sender As Object, e As LinkLabelLinkClickedEventArgs) Handles chklnkBackgroundPicture.LinkClicked
            Try
                Using srcUrlDialog As OpenFileDialog = New OpenFileDialog() With {.Title = "Please Select an image file.", .RestoreDirectory = True, .Filter = "Image Files|*.png;*.bmp;*.gif;*.jpg|All files(*.*)|*.*", .FilterIndex = 0, .Multiselect = False}
                    If srcUrlDialog.ShowDialog() = DialogResult.OK Then
                        _backgroundPictureUrl = srcUrlDialog.FileName
                        pbBackgroundPicture.Image = Image.FromFile(_backgroundPictureUrl)
                        pbBackgroundPicture.ImageLocation = _backgroundPictureUrl
                    End If
                End Using
            Catch
            End Try
        End Sub

        ''' <summary>
        '''
        ''' </summary>
        ''' <returns></returns>
        Public Shadows Function ShowDialog() As DialogResult Implements IDialog.ShowDialog
            Me._tableCellDialog.Element = Nothing
            Return MyBase.ShowDialog()
        End Function

        ''' <summary>
        ''' Handles the Click event of the btnCellProperties control.
        ''' </summary>
        ''' <param name="sender">The source of the event.</param>
        ''' <param name="e">The <see cref="System.EventArgs" /> instance containing the event data.</param>
        Private Sub btnCellProperties_Click(sender As Object, e As EventArgs) Handles btnCellProperties.Click
            If Me._tableCellDialog Is Nothing Then
                MessageBox.Show(
                    "Table Cell Dialog reference is NULL. Please set a TableCellDialog implementing ITableCellDialog to the editor's Dialog.TableCellDialog property.",
                    "Error")
                Return
            End If

            Dim tblCellElement As TableCellElement = Me.getOrInitCellElement()
            If tblCellElement Is Nothing Then
                MessageBox.Show(
                    "No cells were found.",
                    "Error")
                Return
            End If

            Me._tableCellDialog.Element = tblCellElement
            Using Me._tableCellDialog.LockOverrideSettingsToAllCells()
                If Me._tableCellDialog.ShowDialog() = DialogResult.OK Then
                    Me.Element.CellElement = Me._tableCellDialog.Element
                End If
            End Using
        End Sub

        Private Function getOrInitCellElement() As TableCellElement
            Return Me.Element.GetFirstCellElement()
        End Function

        ''' <summary>
        ''' Handles the CheckedChanged event of the chkCaption control.
        ''' </summary>
        ''' <param name="sender">The source of the event.</param>
        ''' <param name="e">The <see cref="System.EventArgs" /> instance containing the event data.</param>
        Private Sub chkCaption_CheckedChanged(sender As Object, e As EventArgs) Handles chkCaption.CheckedChanged
            txtCaption.Enabled = chkCaption.Checked
        End Sub

        ''' <summary>
        ''' Handles the CheckedChanged event of the chkBorderStyle control.
        ''' </summary>
        ''' <param name="sender">The source of the event.</param>
        ''' <param name="e">The <see cref="System.EventArgs" /> instance containing the event data.</param>
        Private Sub chkBorderStyle_CheckedChanged(sender As Object, e As EventArgs) Handles chkBorderStyle.CheckedChanged
            cmbBorderStyle.Enabled = chkBorderStyle.Checked
            chkBorderCollapse.Enabled = chkBorderStyle.Checked
            chklnkBorderColor.Enabled = chkBorderStyle.Checked
            Me.txtBorderColor.Enabled = Me.chkBorderStyle.Checked AndAlso Me.chklnkBorderColor.Checked
            If Not chkBorderStyle.Checked Then chkBorderToAll.Checked = True
        End Sub

        Private Sub numBorderWidth_ValueChanged(sender As Object, e As EventArgs) Handles numBorderWidth.ValueChanged
            If Me.numBorderWidth.Value = 0.0D Then
                Me.chkBorderToAll.Checked = True
            End If
        End Sub

#End Region

        ' IDisposable.Dispose is already satisfied by the WinForms Form base
        ' class (which implements IDisposable), so no bridge is needed here.
        ' (ShowDialog is implemented above as the literal "new" shadow that
        ' also fulfills IDialog.ShowDialog.)
    End Class

End Namespace

File: 09-CustomDialog-CS/Dialogs/TablePropertiesDialog.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Globalization;
using System.Windows.Forms;
using SpiceLogic.HtmlEditor.Abstractions.Entities;
using SpiceLogic.HtmlEditor.WinForms.Helpers;
using SpiceLogic.HtmlEditor.WinForms.Models.Dialogs;

namespace CustomDialog.Dialogs;

/// <summary>
/// Class TablePropertiesDialog
/// </summary>
public partial class TablePropertiesDialog : DialogBase, ITableDialog
{
    /// <summary>
    /// The _element
    /// </summary>
    private TableElement _element;
    /// <summary>
    /// The _table cell dialog
    /// </summary>
    private readonly ITableCellDialog _tableCellDialog;
    /// <summary>
    /// The _background picture URL
    /// </summary>
    private string _backgroundPictureUrl = string.Empty;

    private readonly List<string> _propertiesAffected =
    [
        "Rows",
        "Columns",
        "Width",
        "Height",
        "Caption",
        "BorderWidth",
        "CellPadding",
        "CellSpacing",
        "BorderColor",
        "BorderAttr",
        "BorderStyle",
        "BgColor",
        "BorderCollapse",
        "SummaryDescription",
        "ID",
        "Name",
        "CSSClassName",
        "BackGround"
    ];

    /// <summary>
    /// Initializes a new instance of the <see cref="TablePropertiesDialog" /> class.
    /// </summary>
    /// <param name="cellDialog">The cell dialog.</param>
    public TablePropertiesDialog(ITableCellDialog cellDialog)
    {
        InitializeComponent();
        // Spec 019 FR-009: cmbBorderStyle.SelectedIndex = 0 moved to the
        // .designer.cs file -- design-time serialization, no runtime override.
        this._tableCellDialog = cellDialog;
        this.CancelButton = btnCancel;

        Icon = Constants.FormIcon;
    }

    /// <summary>
    /// Handles the Load event of the TablePropertiesDialog control.
    /// </summary>
    /// <param name="sender">The source of the event.</param>
    /// <param name="e">The <see cref="EventArgs"/> instance containing the event data.</param>
    private void TablePropertiesDialog_Load(object sender, EventArgs e)
    {
        btnCellProperties.Visible = this.getOrInitCellElement() != null;
    }

    /// <summary>
    /// Gets or sets the element.
    /// </summary>
    /// <value>The element.</value>
    public TableElement Element
    {
        get
        {
            if (_element == null)
                _element = new TableElement(null);

            _element.ResetValues(this._propertiesAffected);

            _element.Rows = (int)numRows.Value;
            _element.Columns = (int)numCols.Value;

            if (chkCellPadding.Checked)
                _element.CellPadding = (int)numCellPadding.Value;
            if (chkCellSpacing.Checked)
                _element.CellSpacing = (int)numCellSpacing.Value;

            if (chkWidth.Checked && cmbWidthUnit.SelectedItem != null)
            {
                _element.Width = Convert.ToInt32(txtWidth.Text.Trim());
                _element.WidthUnit = cmbWidthUnit.SelectedItem.ToString();
            }

            if (chkHeight.Checked && cmbHeightUnit.SelectedItem != null)
            {
                _element.Height = Convert.ToInt32(txtHeight.Text.Trim());
                _element.HeightUnit = cmbHeightUnit.SelectedItem.ToString();
            }

            if (chkCaption.Checked)
                _element.Caption = txtCaption.Text.Trim();

            if (chkBorderWidth.Checked)
            {
                if (chkBorderStyle.Checked && cmbBorderStyle.SelectedItem != null)
                {
                    _element.BorderStyle = cmbBorderStyle.SelectedItem.ToString();
                    _element.BorderWidth = (int)numBorderWidth.Value;
                }
                else
                    _element.BorderAttr = (int)numBorderWidth.Value;
            }

            if (chklnkBorderColor.Enabled && chklnkBorderColor.Checked)
                _element.BorderColor = ColorTranslator.ToHtml(txtBorderColor.BackColor);

            if (chklnkBgColor.Checked)
                _element.BgColor = ColorTranslator.ToHtml(txtBgColor.BackColor);

            _element.SummaryDescription = txtSummaryDescription.Text;
            _element.BorderCollapse = chkBorderCollapse.Enabled && chkBorderCollapse.Checked;
            _element.Id = txtId.Text.Trim();
            _element.Name = txtName.Text.Trim();
            _element.CssClassName = txtClassName.Text.Trim();
            _element.CssStyle = txtCss.Text.Trim();

            if (chklnkBackgroundPicture.Checked)
                _element.BackGround = pbBackgroundPicture.ImageLocation;

            _element.BorderToAll = chkBorderToAll.Checked;

            return _element;
        }
        set
        {
            _element = value;
            if (_element == null)
                return;

            // Spec 019 FR-009: every element→UI assignment is gated on the
            // element carrying data, so design-time defaults survive otherwise.
            if (_element.Rows > 0)
                numRows.Value = _element.Rows;
            if (_element.Columns > 0)
                numCols.Value = _element.Columns;

            if (_element.Width.HasValue)
            {
                chkWidth.Checked = true;
                txtWidth.Text = _element.Width.Value.ToString(CultureInfo.InvariantCulture);
                if (_element.WidthUnit != null)
                    cmbWidthUnit.SelectedItem = _element.WidthUnit;
            }

            if (_element.Height.HasValue)
            {
                chkHeight.Checked = true;
                txtHeight.Text = _element.Height.Value.ToString(CultureInfo.InvariantCulture);
                if (_element.HeightUnit != null)
                    cmbHeightUnit.SelectedItem = _element.HeightUnit;
            }

            if (!string.IsNullOrEmpty(_element.Caption))
            {
                chkCaption.Checked = true;
                txtCaption.Text = _element.Caption;
            }

            if (_element.BorderWidth.HasValue || _element.BorderAttr.HasValue)
                chkBorderWidth.Checked = true;

            if (!string.IsNullOrEmpty(_element.BorderStyle))
            {
                chkBorderStyle.Checked = true;
                cmbBorderStyle.Enabled = true;
                chkBorderCollapse.Enabled = true;
                chklnkBorderColor.Enabled = true;
                this.txtBorderColor.Enabled = this.chklnkBorderColor.Checked;
                cmbBorderStyle.Text = _element.BorderStyle;
                if (chkBorderWidth.Checked)
                    numBorderWidth.Value = _element.BorderWidth ?? 0;
            }
            else if (chkBorderWidth.Checked)
            {
                numBorderWidth.Value = _element.BorderAttr ?? (_element.BorderWidth ?? 0);
            }

            if (_element.CellPadding.HasValue)
            {
                chkCellPadding.Checked = true;
                numCellPadding.Value = _element.CellPadding.Value;
            }

            if (_element.CellSpacing.HasValue)
            {
                chkCellSpacing.Checked = true;
                numCellSpacing.Value = _element.CellSpacing.Value;
            }

            if (!string.IsNullOrEmpty(_element.BorderColor))
            {
                chklnkBorderColor.Checked = true;
                ColorConverter converter = new();
                txtBorderColor.BackColor = (Color)converter.ConvertFromString(_element.BorderColor);
            }

            if (!string.IsNullOrEmpty(_element.BgColor))
            {
                chklnkBgColor.Checked = true;
                ColorConverter converter = new();
                txtBgColor.BackColor = (Color)converter.ConvertFromString(_element.BgColor);
            }

            if (_element.BorderCollapse.HasValue && _element.BorderCollapse.Value)
                chkBorderCollapse.Checked = true;
            // Spec 019 FR-009: preserve design-time defaults on empty fields.
            if (!string.IsNullOrEmpty(_element.SummaryDescription))
                txtSummaryDescription.Text = _element.SummaryDescription;
            if (!string.IsNullOrEmpty(_element.Id))
                txtId.Text = _element.Id;
            if (!string.IsNullOrEmpty(_element.Name))
                txtName.Text = _element.Name;
            if (!string.IsNullOrEmpty(_element.CssClassName))
                txtClassName.Text = _element.CssClassName;
            /////////////////
            if (!string.IsNullOrEmpty(_element.BackGround))
            {
                chklnkBackgroundPicture.Checked = true;
                pbBackgroundPicture.ImageLocation = _element.BackGround;
                pbBackgroundPicture.Load();
            }

            string cssText = _element.GetCssStyleWithoutProperties(this._propertiesAffected);
            if (!string.IsNullOrEmpty(cssText))
                txtCss.Text = cssText;
            chkBorderToAll.Checked = true;
        }
    }

    #region ------------ UI Event Handlers --------------

    /// <summary>
    /// Handles the CheckedChanged event of the chkWidth control.
    /// </summary>
    /// <param name="sender">The source of the event.</param>
    /// <param name="e">The <see cref="System.EventArgs" /> instance containing the event data.</param>
    private void chkWidth_CheckedChanged(object sender, EventArgs e)
    {
        txtWidth.Enabled = chkWidth.Checked;
        cmbWidthUnit.Enabled = chkWidth.Checked;
    }

    /// <summary>
    /// Handles the CheckedChanged event of the chkHeight control.
    /// </summary>
    /// <param name="sender">The source of the event.</param>
    /// <param name="e">The <see cref="System.EventArgs" /> instance containing the event data.</param>
    private void chkHeight_CheckedChanged(object sender, EventArgs e)
    {
        txtHeight.Enabled = chkHeight.Checked;
        cmbHeightUnit.Enabled = chkHeight.Checked;
    }

    /// <summary>
    /// Handles the CheckedChanged event of the chkBorderWidth control.
    /// </summary>
    /// <param name="sender">The source of the event.</param>
    /// <param name="e">The <see cref="System.EventArgs" /> instance containing the event data.</param>
    private void chkBorderWidth_CheckedChanged(object sender, EventArgs e)
    {
        numBorderWidth.Enabled = chkBorderWidth.Checked;

        this.chkBorderStyle.Enabled = this.chkBorderWidth.Checked;

        //            chkBorderStyle.Checked = chkBorderWidth.Checked;
        cmbBorderStyle.Enabled = chkBorderWidth.Checked && chkBorderStyle.Checked;
        if (!chkBorderWidth.Checked) chkBorderToAll.Checked = true;
    }

    /// <summary>
    /// Handles the CheckedChanged event of the chkCellPadding control.
    /// </summary>
    /// <param name="sender">The source of the event.</param>
    /// <param name="e">The <see cref="System.EventArgs" /> instance containing the event data.</param>
    private void chkCellPadding_CheckedChanged(object sender, EventArgs e)
    {
        numCellPadding.Enabled = chkCellPadding.Checked;
    }

    /// <summary>
    /// Handles the CheckedChanged event of the chkCellSpacing control.
    /// </summary>
    /// <param name="sender">The source of the event.</param>
    /// <param name="e">The <see cref="System.EventArgs" /> instance containing the event data.</param>
    private void chkCellSpacing_CheckedChanged(object sender, EventArgs e)
    {
        numCellSpacing.Enabled = chkCellSpacing.Checked;
    }

    /// <summary>
    /// Handles the LinkClicked event of the chklnkBorderColor control.
    /// </summary>
    /// <param name="sender">The source of the event.</param>
    /// <param name="e">The <see cref="System.Windows.Forms.LinkLabelLinkClickedEventArgs" /> instance containing the event data.</param>
    private void chklnkBorderColor_LinkClicked(object sender, LinkLabelLinkClickedEventArgs e)
    {
        using ColorDialog myDialog = new()
        {
            AllowFullOpen = true,
            AnyColor = true,
            Color = txtBorderColor.BackColor,
            CustomColors = [ColorTranslator.ToOle(txtBorderColor.BackColor)]
        };
        if (myDialog.ShowDialog() == DialogResult.OK)
            txtBorderColor.BackColor = myDialog.Color;
    }

    /// <summary>
    /// Handles the LinkClicked event of the chklnkBgColor control.
    /// </summary>
    /// <param name="sender">The source of the event.</param>
    /// <param name="e">The <see cref="System.Windows.Forms.LinkLabelLinkClickedEventArgs" /> instance containing the event data.</param>
    private void chklnkBgColor_LinkClicked(object sender, LinkLabelLinkClickedEventArgs e)
    {
        using ColorDialog myDialog = new()
        {
            AllowFullOpen = true,
            AnyColor = true,
            Color = txtBgColor.BackColor,
            CustomColors = [ColorTranslator.ToOle(txtBgColor.BackColor)]
        };
        if (myDialog.ShowDialog() == DialogResult.OK)
            txtBgColor.BackColor = myDialog.Color;
    }

    /// <summary>
    /// Handles the LinkClicked event of the chklnkBackgroundPicture control.
    /// </summary>
    /// <param name="sender">The source of the event.</param>
    /// <param name="e">The <see cref="System.Windows.Forms.LinkLabelLinkClickedEventArgs" /> instance containing the event data.</param>
    private void chklnkBackgroundPicture_LinkClicked(object sender, LinkLabelLinkClickedEventArgs e)
    {
        try
        {
            using OpenFileDialog srcUrlDialog = new()
                { Title = @"Please Select an image file.", RestoreDirectory = true, Filter = @"Image Files|*.png;*.bmp;*.gif;*.jpg|All files(*.*)|*.*", FilterIndex = 0, Multiselect = false };
            if (srcUrlDialog.ShowDialog() == DialogResult.OK)
            {
                _backgroundPictureUrl = srcUrlDialog.FileName;
                pbBackgroundPicture.Image = Image.FromFile(_backgroundPictureUrl);
                pbBackgroundPicture.ImageLocation = _backgroundPictureUrl;
            }
        }
        catch
        {
        }
    }

    /// <summary>
    /// 
    /// </summary>
    /// <returns></returns>
    public new DialogResult ShowDialog()
    {
        this._tableCellDialog.Element = null;
        return base.ShowDialog();
    }

    /// <summary>
    /// Handles the Click event of the btnCellProperties control.
    /// </summary>
    /// <param name="sender">The source of the event.</param>
    /// <param name="e">The <see cref="System.EventArgs" /> instance containing the event data.</param>
    private void btnCellProperties_Click(object sender, EventArgs e)
    {
        if (this._tableCellDialog == null)
        {
            MessageBox.Show(
                "Table Cell Dialog reference is NULL. Please set a TableCellDialog implementing ITableCellDialog to the editor's Dialog.TableCellDialog property.",
                "Error");
            return;
        }

        TableCellElement tblCellElement = this.getOrInitCellElement();
        if (tblCellElement == null)
        {
            MessageBox.Show(
                "No cells were found.",
                "Error");
            return;
        }

        this._tableCellDialog.Element = tblCellElement;
        using (this._tableCellDialog.LockOverrideSettingsToAllCells())
        {
            if (this._tableCellDialog.ShowDialog() == DialogResult.OK)
                this.Element.CellElement = this._tableCellDialog.Element;
        }
    }

    private TableCellElement getOrInitCellElement()
    {
        return this.Element.GetFirstCellElement();
    }

    /// <summary>
    /// Handles the CheckedChanged event of the chkCaption control.
    /// </summary>
    /// <param name="sender">The source of the event.</param>
    /// <param name="e">The <see cref="System.EventArgs" /> instance containing the event data.</param>
    private void chkCaption_CheckedChanged(object sender, EventArgs e)
    {
        txtCaption.Enabled = chkCaption.Checked;
    }

    /// <summary>
    /// Handles the CheckedChanged event of the chkBorderStyle control.
    /// </summary>
    /// <param name="sender">The source of the event.</param>
    /// <param name="e">The <see cref="System.EventArgs" /> instance containing the event data.</param>
    private void chkBorderStyle_CheckedChanged(object sender, EventArgs e)
    {
        cmbBorderStyle.Enabled = chkBorderStyle.Checked;
        chkBorderCollapse.Enabled = chkBorderStyle.Checked;
        chklnkBorderColor.Enabled = chkBorderStyle.Checked;
        this.txtBorderColor.Enabled = this.chkBorderStyle.Checked && this.chklnkBorderColor.Checked;
        if (!chkBorderStyle.Checked) chkBorderToAll.Checked = true;
    }

    private void numBorderWidth_ValueChanged(object sender, EventArgs e)
    {
        if (this.numBorderWidth.Value == 0.0m)
            this.chkBorderToAll.Checked = true;
    }

    #endregion
}